README.md
stylus-spritesmith-brunch
=========================

Adds Stylus support to Brunch with automatic sprite sheet generation.

Usage
-----

Install the plugin via npm with `npm install --save stylus-spritemith-brunch`.

Or, do manual install:

* Add `"stylus-spritesmith-brunch": "x.y.z"` to `package.json` of your Brunch app.
* If you want to use the `master` branch of the plugin, add `"stylus-spritesmith-brunch": "git+ssh://git@github.com:jas/stylus-spritesmith-brunch.git"`.

*Note: This plugin is a replacement for `stylus-brunch`.*

Using Sprites
-------------

This plugin defines Stylus mixins, variables, and functions that make using sprites in your stylesheets almost effortless.

Just drop images into `app/assets/images/sprites`, and a sprite sheet will be compiled to `public/images/spritesheets/default.png`.

If you need to configure the input and output directories, reference the **Plugin Options** section.

### Mixin

The `sprite` mixin is the easiest way to use a sprite image. To use the mixin, you must first import it into your stylesheets.

```
@import "sprites"
```

Then use it like so...

```styl
.logo {
  sprite('logo.png') // finds "icon-star.png" in default sprite sheet
}
```

If you have multiple sprite sheets, a second argument can be given to specify the sprite sheet to which the image belongs.
```
.icon-star {
  sprite('icon-star.png', 'icons') // finds "icon-star.png" in sprite sheet named "icons"
}
```

### Variables and Functions

The following variables and functions are also available for use in your stylesheets.

| Variable               | Type    | Description
|------------------------|---------|-------------
| `sprite-sheets`        | List    | List of sprite sheets generated
| `default-sprite-sheet` | String  | Value of config option `plugins.sprites.defaultSheet`
| `extend-sprite-sheets` | Boolean | Value of config option `plugins.sprites.useExtendDirective`

#### Sprite Sheet Functions

These functions return values about a given `sprite-sheet`.

| Function              | Description
|-----------------------|-------------
| `sprite-sheet-path`   | String — absolute path to the sprite sheet image
| `sprite-sheet-width`  | Unit, `px` — width of an entire sprite sheet
| `sprite-sheet-height` | Unit, `px` — height of an entire sprite sheet

#### Sprite Images

These functions return values about a given `image`. A second, optional `sprite-sheet` argument can be given to specify the sprite sheet to which the image belongs.

| Function              | Description
|-----------------------|-------------
| `sprite-image-width`  | Unit, `px` — width of an individual sprite image
| `sprite-image-height` | Unit, `px` — height of an individual sprite image
| `sprite-image-x`      | Unit, `px` — x-position of a sprite
| `sprite-image-y`      | Unit, `px` — y-position of a sprite

Plugin Options
--------------

Options for Stylus and Spritesmith can be set in the `plugins` object in your `brunch-config`. All values below are the defaults.

### Stylus

```js
plugins: {
  stylus: {
    // Compress CSS output.
    compress: false,

    // Emits debug info that can be used by the FireStylus Firebug plugin.
    firebug: false,

    // Emits comments in the generated CSS indicating the corresponding Stylus lines.
    lineNumbers: false,

    // Include regular CSS on @import.
    includeCss: false
  }
}
```

### Sprites

```js
plugins: {
  sprites: {
    // Use sprite features.
    enabled: true,

    // Directory in which to output sprite sheet images.
    outputDir: 'images/spritesheets',

    // File name of output sprite sheet images.
    outputFile: '{{name}}.png',

    // Include provided `sprite` mixin when compiling Stylus.
    includeMixins: true,

    // Sprite mixin will `@extend` a sprite sheet's styles, instead of
    // including them directly in the selector in which the mixin is used.
    useExtendDirective: false,

    // Name of the default sprite sheet to use when one is not passed to the
    // provided Stylus mixin or functions.
    defaultSheet: 'default',

    // Sprite sheets to generate. The regular expression matches images to
    // include in that particular sprite sheet.
    spriteSheets: {
      'default': /^app\/assets\/images\/sprites\/(.*)\.png$/
    },

    // Engine Spritemith uses to generate images ('phantomjs', 'canvas', 'gm')
    // See https://github.com/Ensighten/spritesmith for more information.
    engine: 'auto',

    // Image packing algoritm ('binary-tree', 'top-down', 'left-right', 'diagonal', 'alt-diagonal')
    algoritm: 'binary-tree',

    // Padding in pixels to add between sprite images
    padding: 0,

    // Options to pass through to engine for settings
    exportOpts: {
      // Format of output sprite sheets ('png', 'jpeg'); Canvas and gm engines only
      format: 'png',

      // Quality of output sprite sheets; gm engine only
      quality: 75,

      // Milliseconds to wait until terminating PhantomJS script; phantomjs engine only
      timeout: 10000
    },

    // Options to pass through to engine for export
    engineOpts: {},
  }
}
```

License
-------

The MIT License (MIT)

Copyright (c) 2013 Jason Sandmeyer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
0.1.4 / 2013-12-17
==================

 * Add `getDependencies`

0.1.3 / 2013-12-13
==================

 * Add keywords to package.json
 * Fix header levels in README

0.1.2 / 2013-12-13
==================

 * Update README with usage, mixin/variables/functions, options

0.1.1 / 2013-12-13
==================

 * Add changelog
 * Make LICENSE a Markdown file
 * Use sprite Stylus functions even when there are no sprite sheets

0.1.0 / 2013-12-13
==================

 * Initial commit


lib/sprite_functions.js
/**
 * Define variables and functions to use within Stylus files to return
 * information about the generated sprite sheets and images.
 */

module.exports = function(context) {
  var self    = context,
      stylus  = self.stylus,
      nodes   = stylus.nodes,
      utils   = stylus.utils,
      results = self.spriteResults;

  var spriteSheets = Object.keys(results);

  function getSpriteSheet(spriteSheet) {
    spriteSheet = spriteSheet || new nodes.Null();

    if (spriteSheet.toBoolean().isFalse) {
      spriteSheet = new nodes.String(self.defaultSheet);
    }

    utils.assertType(spriteSheet, 'string', 'spriteSheet');

    if (!results[spriteSheet.val]) {
      var err = new Error('Could not find sprite sheet "' + spriteSheet.val + '"');
      throw utils.formatException(err, spriteSheet);
    }

    return results[spriteSheet.val];
  }

  function spriteSheetPath() {
    return function(spriteSheet) {
      var data = getSpriteSheet(spriteSheet);
      return new nodes.String(data.url);
    }
  }

  function spriteSheetProperty(property) {
    return function(spriteSheet) {
      var data = getSpriteSheet(spriteSheet);
      return new nodes.Unit(data.properties[property], 'px');
    }
  }

  function spriteImageProperty(property) {
    return function(image, spriteSheet) {
      var data = getSpriteSheet(spriteSheet);

      utils.assertType(image, 'string', 'image');

      if (!data.coordinates[image.val]) {
        var err = new Error('Could not find image "' + image.val + '" in sprite sheet "' + spriteSheet.val + '"');
        throw utils.formatException(err, spriteSheet);
      }

      return new nodes.Unit(data.coordinates[image.val][property], 'px');
    }
  }

  return function(compiler) {
    // Placeholder Selectors and Mixins
    if (self.includeMixins) {
      compiler.include(__dirname);
    }

    // Variables
    compiler.define('sprite-sheets', spriteSheets);
    compiler.define('default-sprite-sheet', self.defaultSheet);
    compiler.define('extend-sprite-sheets', self.useExtendDirective);

    // Functions
    compiler.define('sprite-sheet-path', spriteSheetPath());
    compiler.define('sprite-sheet-width', spriteSheetProperty('width'));
    compiler.define('sprite-sheet-height', spriteSheetProperty('height'));
    compiler.define('sprite-image-width', spriteImageProperty('width'));
    compiler.define('sprite-image-height', spriteImageProperty('height'));
    compiler.define('sprite-image-x', spriteImageProperty('x'));
    compiler.define('sprite-image-y', spriteImageProperty('y'));
  }
};


lib/index.js
var logger      = require('loggy'),
    path        = require('path'),
    fs          = require('fs'),
    async       = require('async'),
    progeny     = require('progeny'),
    find        = require('findit'),
    mkdirp      = require('mkdirp'),
    stylus      = require('stylus'),
    spritesmith = require('spritesmith'),
    nib         = require('nib'),
    spriteFns   = require('./sprite_functions');

/**
 * Stylus Compiler with Spritesmith
 *
 * Generates sprite sheets from series of images, defines sprite variables
 * and functions for use in Stylus, and compiles stylesheets to CSS.
 */
function StylusCompiler(config) {
  var plugins = config.plugins;

  // Config
  this.config = config;
  this.stylusConfig = (plugins && plugins.stylus) || {};
  this.spriteConfig = (plugins && plugins.sprites) || {};

  // Modules
  this.stylus = this.stylusConfig.module || stylus;
  this.spritesmith = this.spriteConfig.module || spritesmith;

  // Is sprite generation enabled?
  this.spritesEnabled = this.spriteConfig.enabled === undefined ? true : this.spriteConfig.enabled;

  // Collection of sprite sheets to generate
  this.spriteSheets = this.spriteConfig.spriteSheets || {
    'default': /^app\/assets\/images\/sprites\/(.*)\.png$/
  };

  // Directory in which to output sprite sheets
  this.outputDir = this.spriteConfig.outputDir || 'images/spritesheets';

  // File name to give to output sprite sheets
  this.outputFile = this.spriteConfig.outputFile || '{{name}}.png';

  // Include sprite placeholder selectors and mixins
  this.includeMixins = this.spriteConfig.includeMixins === undefined ? true : this.spriteConfig.includeMixins;

  // Each sprite will `@extend` a sprite sheet class instead of mixing in properties
  this.useExtendDirective = !!this.spriteConfig.useExtendDirective;

  // Default stylesheet when one is not specified in Stylus `sprite` function
  this.defaultSheet = this.spriteConfig.defaultSheet || 'default';

  // Directories to ignore while walking the project for sprites
  this.ignoredDirs = ['.git', 'node_modules', 'bower_components', 'vendor'];

  // Resulting generated sprite data
  this.spriteResults = {};

  // Flags
  this.generatingSprites = false;
  this.spritesCreated = false;

  // Get dependencies
  this.getDependencies = progeny({
    rootPath: this.config.paths.root
  });
}

StylusCompiler.prototype.brunchPlugin = true;
StylusCompiler.prototype.type = 'stylesheet';
StylusCompiler.prototype.extension = 'styl';

/**
 * Invoked by Brunch before compilation of each stylesheet file. Places the
 * arguments passed to the function into a queue while the sprite sheets are
 * prepared and generated once. Afterwards, each callback and Brunch proceeds
 * to `compile`.
 */
StylusCompiler.prototype.lint = function buildSprites(data, filePath, callback) {
  var self = this;

  self.queue = self.queue || [];

  // Proceed to `compile` if sprite sheets are disabled or already generated
  if (!this.spritesEnabled || this.spritesCreated) {
    return callback();
  }

  // Start spritesheet generation if it has not already started or completed
  if (!this.generatingSprites && !this.spritesCreated) {
    self.findSprites(function(sprites) {
      self.generateSprites(sprites, function() {
        self.queue.forEach(function(args) {
          var callback = args[2];
          callback();
        });
      });
    });

  }

  if (!this.spritesCreated) {
    self.queue.push([data, filePath, callback]);
  } else {
    callback();
  }
};

/**
 * Compile Stylus files to CSS with given options.
 */
StylusCompiler.prototype.compile = function compileStylus(data, filePath, callback) {
  var self = this,
      config = this.stylusConfig,
      compiler = this.stylus(data);

  // Set options on the compiler
  compiler
    .set('filename', filePath)
    .set('compress', !!config.compress)
    .set('firebug', !!config.firebug)
    .set('linenos', !!config.lineNumbers)
    .set('include css', !!config.includeCss)

  // Use sprite variables and functions
  if (this.spritesEnabled) {
    compiler.use(spriteFns(this));
  }

  // Use any additional plugins
  compiler.use(nib());

  return compiler.render(callback);
};

/**
 * Walk project directories to find sprite images that match the regular
 * expressions set for each sprite in `this.spriteSheets`.
 */
StylusCompiler.prototype.findSprites = function(callback) {
  var self = this,
      spriteSheets = {},
      finder;

  this.generatingSprites = true;

  // Start walking from the Brunch project root
  finder = find(this.config.paths.root);

  // Stop walking a directory if it's in `this.ignoredDirs`
  finder.on('directory', function(dir, stat, stop) {
    if (self.ignoredDirs.indexOf(dir) > -1) {
      return stop();
    }
  });

  // Test each sprite sheet's regex against each file
  finder.on('file', function(filePath, stat) {
    Object.keys(self.spriteSheets).forEach(function(name) {
      var regex = self.spriteSheets[name];

      if (regex.test(filePath)) {
        spriteSheets[name] = spriteSheets[name] || [];
        spriteSheets[name].push(filePath);
      }
    });
  });

  // Callback with the sprite file paths
  finder.on('end', function() {
    callback(spriteSheets);
  });
};

/**
 * Generate sprite images, coordinates, and properties.
 */
StylusCompiler.prototype.generateSprites = function(spriteSheets, callback) {
  var self = this,
      config = this.spriteConfig,
      publicDir = this.config.paths.public;

  // Generate each sprite sheet
  async.each(Object.keys(spriteSheets), function(spriteSheet, callback) {
    var outputDir = path.join(publicDir, self.outputDir),
        outputFile = self.outputFile.replace(/\{\{name\}\}/, spriteSheet),
        outputPath = path.join(outputDir, outputFile),
        cssPath = '/' + self.outputDir + '/' + outputFile;

    // Spritesmith options
    var options = {
      src: spriteSheets[spriteSheet],
      engine: config.engine || 'auto',
      algorithm: config.algorith || 'binary-tree',
      padding: config.padding,
      exportOpts: config.exportOpts || {},
      engineOpts: config.engineOpts || {}
    };

    options.exportOpts = options.exportOpts.format || 'png';
    options.exportOpts = options.exportOpts.quality || 75;
    options.exportOpts = options.exportOpts.timeout || 10000;

    self.spritesmith(options, function(err, result) {
      if (err) {
        logger.error('failed to generate sprite sheet "' + outputPath + '"');
        return callback(err);
      }

      result.url = cssPath;

      // Store sprite data for use in Stylus functions
      renameCoordinatesKeys(result.coordinates, function(coordinates) {
        result.coordinates = coordinates;
        self.spriteResults[spriteSheet] = result;
      });

      // Write image to disk
      mkdirp(outputDir, function(err) {
        fs.writeFile(outputPath, result.image, 'binary', function(err) {
          if (err) {
            logger.error('failed to write sprite sheet "' + outputPath + '" to disk');
            return callback(err);
          }

          logger.info('generated sprite sheet "' + outputPath + '"');
          callback();
        });
      });
    });
  }, function(err) {
    if (err) return callback(err);

    self.generatingSprites = false;
    self.spritesCreated = true;

    return callback();
  });
};

/**
 * Rename coordinate keys from full image path to just the filename.
 */
function renameCoordinatesKeys(coordinates, callback) {
  var newCoordinates = {};

  Object.keys(coordinates).forEach(function(spriteSheet) {
    newCoordinates[path.basename(spriteSheet)] = coordinates[spriteSheet];
  });

  callback(newCoordinates);
};

module.exports = StylusCompiler;
